packages/astro/src/assets/build/generate.ts

```diff
@@ -0,0 +1,174 @@
+import fs, { readFileSync } from 'node:fs';
+import { basename, join } from 'node:path/posix';
+import type { StaticBuildOptions } from '../../core/build/types.js';
+import { warn } from '../../core/logger/core.js';
+import { prependForwardSlash } from '../../core/path.js';
+import { isServerLikeOutput } from '../../prerender/utils.js';
+import { getConfiguredImageService, isESMImportedImage } from '../internal.js';
+import type { LocalImageService } from '../services/service.js';
+import type { ImageMetadata, ImageTransform } from '../types.js';
+import { loadRemoteImage, type RemoteCacheEntry } from './remote.js';
+
+interface GenerationDataUncached {
+	cached: false;
+	weight: {
+		before: number;
+		after: number;
+	};
+}
+
+interface GenerationDataCached {
+	cached: true;
+}
+
+type GenerationData = GenerationDataUncached | GenerationDataCached;
+
+export async function generateImage(
+	buildOpts: StaticBuildOptions,
+	options: ImageTransform,
+	filepath: string
+): Promise<GenerationData | undefined> {
+	let useCache = true;
+	const assetsCacheDir = new URL('assets/', buildOpts.settings.config.cacheDir);
+
+	// Ensure that the cache directory exists
+	try {
+		await fs.promises.mkdir(assetsCacheDir, { recursive: true });
+	} catch (err) {
+		warn(
+			buildOpts.logging,
+			'astro:assets',
+			`An error was encountered while creating the cache directory. Proceeding without caching. Error: ${err}`
+		);
+		useCache = false;
+	}
+
+	let serverRoot: URL, clientRoot: URL;
+	if (isServerLikeOutput(buildOpts.settings.config)) {
+		serverRoot = buildOpts.settings.config.build.server;
+		clientRoot = buildOpts.settings.config.build.client;
+	} else {
+		serverRoot = buildOpts.settings.config.outDir;
+		clientRoot = buildOpts.settings.config.outDir;
+	}
+
+	const isLocalImage = isESMImportedImage(options.src);
+
+	const finalFileURL = new URL('.' + filepath, clientRoot);
+	const finalFolderURL = new URL('./', finalFileURL);
+
+	// For remote images, instead of saving the image directly, we save a JSON file with the image data and expiration date from the server
+	const cacheFile = basename(filepath) + (isLocalImage ? '' : '.json');
+	const cachedFileURL = new URL(cacheFile, assetsCacheDir);
+
+	await fs.promises.mkdir(finalFolderURL, { recursive: true });
+
+	// Check if we have a cached entry first
+	try {
+		if (isLocalImage) {
+			await fs.promises.copyFile(cachedFileURL, finalFileURL);
+
+			return {
+				cached: true,
+			};
+		} else {
+			const JSONData = JSON.parse(readFileSync(cachedFileURL, 'utf-8')) as RemoteCacheEntry;
+
+			// If the cache entry is not expired, use it
+			if (JSONData.expires < Date.now()) {
```

Comment:
@y-nk sorry to comment on a closed PR, but is this logic correct? I _think_ it should be `>`, but I may be missing something obvious.

(but big ❤️❤️❤️ for this PR overall - it's super helpful to have remote images optimised locally!)

Reply:
@notjosh i'll be honest i don't remember writing that line. maybe @Princesseuh have more clue about it. from reading the if it seems you're right, but i'm not so sure.

Reply:
This was fixed in a PR after this one!